File: app/src/main/java/com/nolamarel/onlinelibrary/Fragments/BooksFragment.java
package com.nolamarel.onlinelibrary.Fragments;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.nolamarel.onlinelibrary.OnItemClickListener;
import com.nolamarel.onlinelibrary.R;
import com.nolamarel.onlinelibrary.Adapters.books.Book;
import com.nolamarel.onlinelibrary.Adapters.books.BookAdapter;
import com.nolamarel.onlinelibrary.databinding.FragmentBooksBinding;

import java.util.ArrayList;

public class BooksFragment extends Fragment{
    private FragmentBooksBinding binding;
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentBooksBinding.inflate(inflater, container, false);

        binding.arrowBack.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getParentFragmentManager().popBackStack();
            }
        });

        String sectionId = getArguments().getString("sectionId");
        loadBooks(sectionId);

        return binding.getRoot();
    }

    private void loadBooks(String sectionId){
        ArrayList<Book> books = new ArrayList<>();

        FirebaseDatabase.getInstance().getReference().addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {

                if(snapshot.child("genres").child(sectionId).child("books").getValue() != null){
                    binding.sectionName.setText(snapshot.child("genres").child(sectionId).child("name").getValue().toString());
                    String booksStr = snapshot.child("genres").child(sectionId).child("books").getValue().toString();
                    String[] booksIds = booksStr.split(",");

                    for (String bookId : booksIds){
                        DataSnapshot bookSnapshot = snapshot.child("Books").child(bookId);

                        String bookName = bookSnapshot.child("name").getValue().toString();
                        String bookAuthor = bookSnapshot.child("author").getValue().toString();
                        String myBookId = bookSnapshot.getKey().toString();
                        String bookImage = bookSnapshot.child("image").getValue().toString();


                        books.add(new Book(myBookId, bookAuthor, bookName, bookImage));

                    }

                    binding.booksRv.setLayoutManager(new LinearLayoutManager(getContext()));
                binding.booksRv.setAdapter(new BookAdapter(books, new OnItemClickListener.ItemClickListener() {
                    @Override
                    public void onItemClick(int position) {
                        String selectedBookId = books.get(position).getBookId();
                        Fragment fragment = new BookDescriptionFragment();
                        Bundle args = new Bundle();
                        args.putString("bookId", selectedBookId);

                        fragment.setArguments(args);
                        getParentFragmentManager().beginTransaction().replace(R.id.fragment_container, fragment).addToBackStack(null).commit();
                    }
                }));

                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }
}

File: app/src/main/java/com/nolamarel/onlinelibrary/Fragments/BooksFragment.kt
package com.nolamarel.onlinelibrary.Fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.nolamarel.onlinelibrary.Adapters.books.Book
import com.nolamarel.onlinelibrary.Adapters.books.BookAdapter
import com.nolamarel.onlinelibrary.ApiClient
import com.nolamarel.onlinelibrary.OnItemClickListener
import com.nolamarel.onlinelibrary.R
import com.nolamarel.onlinelibrary.databinding.FragmentBooksBinding
import kotlinx.coroutines.launch

class BooksFragment : Fragment() {
    private lateinit var binding: FragmentBooksBinding

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentBooksBinding.inflate(inflater, container, false)

        binding.arrowBack.setOnClickListener {
            parentFragmentManager.popBackStack()
        }

        val sectionId = arguments?.getString("sectionId") ?: ""
        val sectionName = arguments?.getString("sectionName") ?: ""

        loadBooks(sectionId, sectionName)

        return binding.root
    }

    private fun loadBooks(sectionId: String, sectionName: String) {
        viewLifecycleOwner.lifecycleScope.launch {
            try {
                val response = ApiClient.serverApi.getBooksByGenre(sectionId)

                if (response.isSuccessful) {
                    val bookDTOs = response.body().orEmpty()
                    val books = ArrayList(bookDTOs.map {
                        Book(
                            bookId = it.bookId.toString(),
                            bookAuthor = it.author,
                            bookName = it.title,
                            bookImage = it.coverUrl
                        )
                    })

                    binding.sectionName.text = sectionName
                    binding.booksRv.layoutManager = LinearLayoutManager(context)
                    binding.booksRv.adapter = BookAdapter(
                        books,
                        object : OnItemClickListener.ItemClickListener {
                            override fun onItemClick(position: Int) {
                                val selectedBookId = books[position].bookId
                                val fragment = BookDescriptionFragment().apply {
                                    arguments = Bundle().apply {
                                        putString("bookId", selectedBookId)
                                    }
                                }
                                parentFragmentManager.beginTransaction()
                                    .replace(R.id.fragment_container, fragment)
                                    .addToBackStack(null)
                                    .commit()
                            }
                        }
                    )
                } else {
                    Toast.makeText(context, "Ошибка загрузки книг", Toast.LENGTH_SHORT).show()
                }
            } catch (e: Exception) {
                e.printStackTrace()
                Toast.makeText(context, "Ошибка подключения к серверу", Toast.LENGTH_SHORT).show()
            }
        }
    }
}